### ISSUE TITLE
`Eio.Net.connect` cannot be cancelled when using `luv` backend

### ISSUE THREAD

**Issue by nojb** [ocaml-multicore/eio#303]
What the title says. Tried with `Eio.Time.with_timeout_exn clock 3.0 (fun () -> Eio.Net.connect ~sw net stream)`. `io-uring` worked as expected.

**Comment by haesbaert** [comment]
On a preliminary look it seems it's not really supported by Luv:
`http://docs.libuv.org/en/v1.x/request.html#c.uv_cancel`

We also have this comment in `eio_luv.ml`:
```
  (* todo: how do you cancel connect operations with luv? *)
  method connect ~sw = function
    | `Tcp (host, port) ->
      let sock = Luv.TCP.init ~loop:(get_loop ()) () |> or_raise |> Handle.of_luv ~sw in
      let addr = luv_addr_of_eio host port in
      await_exn (fun _loop _fiber -> Luv.TCP.connect (Handle.get "connect" sock) addr);
      (socket sock :> < Eio.Flow.two_way; Eio.Flow.close> )
    | `Unix path ->
      let sock = Luv.Pipe.init ~loop:(get_loop ()) () |> or_raise |> Handle.of_luv ~sw in
      await_exn (fun _loop _fiber -> Luv.Pipe.connect (Handle.get "connect" sock) path);
      (socket sock :> < Eio.Flow.two_way; Eio.Flow.close> )
```

**Comment by nojb** [comment]
Yes, I had seen the comment. Can this be documented somehow?

**Comment by talex5** [comment]
Maybe calling `shutdown` on it would work?

**Comment by haesbaert** [comment]
> Maybe calling `shutdown` on it would work?

It does not, tried all shutdown/close.
```ocaml
open Eio

let addr = `Tcp ((Eio_unix.Ipaddr.of_unix (Unix.inet_addr_of_string "142.250.186.174")), 8080)

let test env =
  let net = Stdenv.net env in
  let clock = Stdenv.clock env in
  Switch.run @@ fun sw ->
  Time.with_timeout_exn clock 1.0 @@ fun () ->
  let flow = Net.connect ~sw net addr in
  Flow.copy_string "Hi there" flow;
  Flow.shutdown flow `Send;
  Flow.shutdown flow `Receive;
  ()

let () =
  let () = try Eio_linux.run test with Time.Timeout -> print_endline "Uring terminates" in
  let () = try Eio_luv.run test with Time.Timeout -> print_endline "Luv terminates" in
  ()
```

Lubuv does this: https://github.com/libuv/libuv/blob/v1.x/src/unix/tcp.c#L205
and this:
https://github.com/libuv/libuv/blob/v1.x/src/unix/stream.c#L1264

Stevens says the file descriptor should become writable and readable in the case of an "error", so libuv should be waking up, I need more time to figure libuv.

**Comment by talex5** [comment]
>> Maybe calling shutdown on it would work?
>
> It does not, tried all shutdown/close.

I meant something like this:

```ocaml
  method connect ~sw = function
    | `Tcp (host, port) ->
      let sock = Luv.TCP.init ~loop:(get_loop ()) () |> or_raise |> Handle.of_luv ~sw in
      let addr = luv_addr_of_eio host port in
      enter (fun st k ->
          let cancel_reason = ref None in
          Luv.TCP.connect (Handle.get "connect" sock) addr (fun v ->
            if Fiber_context.clear_cancel_fn k.fiber then (
              enqueue_thread st k v
            ) else (
              (* Cancellations always come from the same domain, so we can be sure
                 that [cancel_reason] is set by now. *)
              enqueue_failed_thread st k (Option.get !cancel_reason)
            )
          );
          Fiber_context.set_cancel_fn k.fiber (fun ex ->
              cancel_reason := Some ex;
              match Handle.to_unix_opt `Peek sock with
              | Some fd -> Unix.shutdown fd Unix.SHUTDOWN_ALL
              | None -> ()
            );
        )
      |> or_raise;
      (socket sock :> < Eio.Flow.two_way; Eio.Flow.close> )
    | `Unix path ->
      let sock = Luv.Pipe.init ~loop:(get_loop ()) () |> or_raise |> Handle.of_luv ~sw in
      await_exn (fun _loop _fiber -> Luv.Pipe.connect (Handle.get "connect" sock) path);
      (socket sock :> < Eio.Flow.two_way; Eio.Flow.close> )
```

With that, I get:
```
$ dune exec -- ./test.exe
Uring terminates                      
Luv terminates
```

**Comment by haesbaert** [comment]
Ahhh I see, my code was nonsensical.
I see how this terminates the Fiber, what I still don't get is: how is the handle removed from Libuv in this case ?
On connect the Libuv ends up linking the handle inside `loop `and adding to one of its lists, how does it get removed ?

**Comment by talex5** [comment]
> I see how this terminates the Fiber, what I still don't get is: how is the handle removed from Libuv in this case ?

Yep, that was missing. It's fixed in #311.
